neqo-http3/src/hframe.rs

```diff
@@ -37,14 +38,116 @@ pub enum HStreamType {
     Push,
 }
 
-#[derive(PartialEq, Debug)]
+#[derive(Clone, PartialEq, Debug, Copy)]
 pub enum HSettingType {
     MaxHeaderListSize,
-    MaxTableSize,
+    MaxTableCapacity,
     BlockedStreams,
     UnknownType,
 }
 
+fn hsetting_default(setting_type: HSettingType) -> u64 {
+    match setting_type {
+        HSettingType::MaxHeaderListSize => 1 << 62,
+        _ => 0,
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct HSetting {
+    pub setting_type: HSettingType,
+    pub value: u64,
+}
+
+impl HSetting {
+    pub fn new(setting_type: HSettingType, value: u64) -> Self {
+        HSetting {
+            setting_type,
+            value,
+        }
+    }
+}
+
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct HSettings {
+    settings: Vec<HSetting>,
+}
+
+impl HSettings {
+    pub fn new(settings: Vec<HSetting>) -> Self {
+        HSettings { settings }
+    }
+
+    pub fn get(&self, setting: HSettingType) -> u64 {
+        match self.settings.iter().find(|s| s.setting_type == setting) {
+            Some(v) => v.value,
+            None => hsetting_default(setting),
+        }
+    }
+
+    pub fn encode_settings(&self, enc: &mut Encoder) {
+        for iter in self.settings.iter() {
+            match iter.setting_type {
+                HSettingType::MaxHeaderListSize => {
+                    enc.encode_varint(SETTINGS_MAX_HEADER_LIST_SIZE as u64);
+                    enc.encode_varint(iter.value);
+                }
+                HSettingType::MaxTableCapacity => {
+                    enc.encode_varint(SETTINGS_QPACK_MAX_TABLE_CAPACITY as u64);
+                    enc.encode_varint(iter.value);
+                }
+                HSettingType::BlockedStreams => {
+                    enc.encode_varint(SETTINGS_QPACK_BLOCKED_STREAMS as u64);
+                    enc.encode_varint(iter.value);
+                }
+                HSettingType::UnknownType => {}
+            }
+        }
+    }
+
+    pub fn decode_settings(&mut self, dec: &mut Decoder) -> Res<()> {
+        while dec.remaining() > 0 {
+            let st_read = match dec.decode_varint() {
+                Some(v) => v,
+                _ => return Err(Error::NotEnoughData),
+            };
+            let setting_type = match st_read {
+                SETTINGS_MAX_HEADER_LIST_SIZE => HSettingType::MaxHeaderListSize,
+                SETTINGS_QPACK_MAX_TABLE_CAPACITY => HSettingType::MaxTableCapacity,
+                SETTINGS_QPACK_BLOCKED_STREAMS => HSettingType::BlockedStreams,
+                _ => HSettingType::UnknownType,
+            };
+            let value = match dec.decode_varint() {
+                Some(v) => v,
+                _ => return Err(Error::NotEnoughData),
+            };
+            if setting_type != HSettingType::UnknownType {
+                self.settings.push(HSetting::new(setting_type, value));
+            }
+        }
+        Ok(())
+    }
+}
+
+impl From<Vec<HSetting>> for HSettings {
```

Comment:
You should implement conversion from a slice reference (`From<&[HSetting]>`) instead.  But how often do you really need this?